File: config/src/main/scala/net/kemitix/thorp/config/ConfigValidation.scala
package net.kemitix.thorp.config

import java.io.File

sealed trait ConfigValidation {

  def errorMessage: String
}

object ConfigValidation {

  case object SourceIsNotADirectory extends ConfigValidation {
    override def errorMessage: String = "Source must be a directory"
  }

  case object SourceIsNotReadable extends ConfigValidation {
    override def errorMessage: String = "Source must be readable"
  }

  case object BucketNameIsMissing extends ConfigValidation {
    override def errorMessage: String = "Bucket name is missing"
  }

  final case class ErrorReadingFile(
      file: File,
      message: String
  ) extends ConfigValidation {
    override def errorMessage: String = s"Error reading file '$file': $message"
  }

}

File: config/src/main/scala/net/kemitix/thorp/config/ConfigValidation.java
package net.kemitix.thorp.config;

import java.io.File;

@FunctionalInterface
public interface ConfigValidation {
    String errorMessage();

    static ConfigValidation sourceIsNotADirectory() {
        return () -> "Source must be a directory";
    }

    static ConfigValidation sourceIsNotReadable() {
        return () -> "Source must be readable";
    }

    static ConfigValidation bucketNameIsMissing() {
        return () -> "Bucket name is missing";
    }

    static ConfigValidation errorReadingFile(File file, String message) {
        return () -> String.format(
                "Error reading file '%s': %s",
                file, message);
    }
}
